flink-python/src/main/java/org/apache/flink/python/util/PythonHelperScriptUtils.java

```diff
@@ -0,0 +1,179 @@
+
+	private static final String CHECK_PIP_VERSION_SCRIPT =
+		"import sys;" +
+		"from pkg_resources import get_distribution, parse_version;" +
+		"pip_version = get_distribution('pip').version;" +
+		"print(parse_version(pip_version) >= parse_version(sys.argv[1]))";
+
+	private static final String GET_RUNNER_SCRIPT_DIR =
+		"import pyflink;" +
+		"import os;" +
+		"print(os.path.join(os.path.abspath(os.path.dirname(pyflink.__file__)), 'bin'))";
+
+	public static void pipInstallRequirements(
+			String requirementsFilePath,
+			@Nullable String requirementsCacheDir,
+			String requirementsInstallDir,
+			String pythonExecutable,
+			Map<String, String> environmentVariables) throws IOException, InterruptedException {
+		String sitePackagesPath = getSitePackagesPath(requirementsInstallDir, pythonExecutable, environmentVariables);
+		String path = String.join(File.pathSeparator, requirementsInstallDir, "bin");
+		appendToEnvironmentVariable("PYTHONPATH", sitePackagesPath, environmentVariables);
+		appendToEnvironmentVariable("PATH", path, environmentVariables);
+
+		List<String> commands = new ArrayList<>(Arrays.asList(
+			pythonExecutable, "-m", "pip", "install", "--ignore-installed", "-r", requirementsFilePath));
+		if (isPipVersionNotLowerThan("8.0.0", pythonExecutable, environmentVariables)) {
+			commands.addAll(Arrays.asList("--prefix", requirementsInstallDir));
+		} else {
+			commands.addAll(Arrays.asList("--install-option", "--prefix=" + requirementsInstallDir));
+		}
+		if (requirementsCacheDir != null) {
+			commands.addAll(Arrays.asList("--find-links", requirementsCacheDir));
+		}
+
+		int retries = 0;
+		while (true) {
+			try {
+				executeHelperScript(commands.toArray(new String[0]), environmentVariables);
+				break;
+			} catch (Throwable t) {
+				retries++;
+				if (retries < MAX_RETRY_TIMES) {
+					LOG.warn(String.format("Pip install failed, retrying... (%d/%d)", retries, MAX_RETRY_TIMES), t);
+				} else {
+					LOG.error(String.format("Pip install failed, already retried %d time...", retries));
+					throw t;
+				}
+			}
+		}
+	}
+
+	public static String getRunnerScriptPath(
+			String pythonExecutable,
+			Map<String, String> environmentVariables) throws IOException, InterruptedException {
+		String[] commands = new String[] { pythonExecutable, "-c", GET_RUNNER_SCRIPT_DIR };
+		String out = executeHelperScript(commands, environmentVariables);
+		String runnerScriptPath;
+		if (OperatingSystem.isWindows()) {
+			runnerScriptPath = String.join(File.separator, out.trim(), PYFLINK_UDF_RUNNER_BAT);
+		} else {
+			runnerScriptPath = String.join(File.separator, out.trim(), PYFLINK_UDF_RUNNER_SH);
+		}
+		if (!new File(runnerScriptPath).exists()) {
+			throw new FileNotFoundException(String.format(
+				"The runner script '%s' does not exist! " +
+				"Please reinstall the apache-flink Python package.", runnerScriptPath));
+		}
+		return runnerScriptPath;
+	}
+
+	private static String getSitePackagesPath(
```

Comment:
Why need to add the site package path to `PYTHONPATH` manually?

Reply:
Because this "site-packages" is a temporary directory, which would not be placed to the `PYTHONPATH` by default.